### DISCUSSION TITLE
What's the best way to provide C functions that are common but not on z/OS?

### DISCUSSION THREAD

**Discussion by MikeFultonDev** [zopencommunity/meta#208] [Reference] [answered]
There are a few functions that I've tripped over in porting that are really just 'code' and don't require any fancy underlying OS support. The latest one I've hit (not for the first time) is the `getopts` function. 

Many of the ports pull in gnulib (at the source level) which lets us get the function cleanly, and for those ports there is no issue.

Some ports just _expect_ the function to be there. The latest is htopport which expects getoptlong.

There are  a few ways to tackle this:
- work with the community to add gnulib to a particular port (like htop) to get gnulib added. This is probably the 'best' technically but may not succeed if the community is not interested
- create a `zosgnulibport` library that pulls in gnulib functions that we need on z/OS and creates an archive that can be bound with other code
- integrate gnulib into zoslib library so that people that use zoslib (most ports) get it 'for free'. The downside of this one is that the zoslib port would need to change it's license to GPL I think
- others I haven't thought of

What do people think?

**Comment by AnthonyGiorgio** [comment]
What about looking to the BSD community for these sort of functions?  Their license is much more friendly to this sort of thing.

https://cgit.freebsd.org/src/tree/lib/libc/stdlib/getopt_long.c

**Comment by perry-ca** [comment, marked as answer]
The zoslib library has been doing this.  I say we continue to expand the coverage in zoslib as needed.  Any code added to zoslib should be licence free or at worst be Apache.  If we add GPL code to zoslib then we run into a problem of IBM products using zoslib.  We need to be careful with the licence of any code added to the lib.

**Comment by IgorTodorovskiIBM** [comment]
I'm slightly biased, but I agree with the license concerns with gnulib. That may very well be why some projects do not use gnulib, due to it's less permissive license. 

Libbsd is interesting, and the license is more open. Actually we used libbsd in the z/OS Swift port a few years back with success. @ccw-1 was behind that. 
I'm open to a hybrid solution, with perhaps zoslib leveraging libbsd?

**Reply by ccw-1** [reply]
zoslib originated from the work we did for swift runtime, then node runtime. It doesn't include functions from the original libbsd. But those functions are fairly easy to do. Except for those that are impossible to do. (like setproctitle). I suggest we just add whatever is needed into zoslib.

**Comment by MikeFultonDev** [comment]
We have a dependency on zoslib for our tools now so closing this
